### PR TITLE
Add animated profile avatars

### PR DESCRIPTION
**Pull request by klopez4212**
## Summary
- add animated avatar capture, review, framing, background, and poster-frame tools
- save animated avatars as poster-first URLs with hover playback and loading feedback
- add camera permissions, APNG/MediaPipe support, haptics, and focused coverage

## Validation
- ./bin/pnpm --dir desktop check
- ./bin/pnpm --dir desktop typecheck
- ./bin/pnpm --dir desktop test
- ./bin/pnpm --dir desktop exec playwright test animated-avatar-screenshots.spec.ts --project=smoke
- ./bin/pnpm --dir desktop exec playwright test profile.spec.ts --project=integration

### CODE REVIEW

**Review by wpfleger96 — COMMENTED** [review]
## Code Review: Add animated profile avatars (#1031)

### Summary

This PR adds a full animated avatar capture pipeline: camera access, MediaPipe selfie segmentation (background removal), per-frame composition with customizable backdrop disc/framing/outline, APNG encoding (ping-pong loop), poster-frame selection, and a URL scheme that embeds both poster and animation in a single `picture` field for backward-compatible Nostr round-tripping. The feature integrates into both the settings profile card and the onboarding avatar step.

### Architecture & Design

**URL scheme** (`#buzz-anim=<encoded-animation-url>`) — Elegant backward-compatible design. Clients that don't understand the fragment simply load the poster URL (fragments are never sent over HTTP). The `parseAnimatedAvatarUrl` function validates both parts are `http(s)` URLs, rejects malformed percent-encoding, and handles null/undefined/empty gracefully. Tests cover all edge cases including XSS attempts.

**Capture pipeline** (`animatedAvatarCapture.ts`, 718 lines) — Well-structured separation of concerns:
- Camera management (`openAvatarCamera`, `stopAvatarCamera`, `listAvatarCameras`)
- Segmentation (lazy-loaded MediaPipe with CDN-pinned version, graceful fallback when offline)
- Frame processing (dark-detail denoising, temporal mask stabilization, bilinear confidence sampling)
- Composition (backdrop circle + person clip + outline + framing offsets)
- Encoding (APNG via `upng-js` with ping-pong loop)

**Component decomposition** — The capture UI is split into focused components:
- `AnimatedAvatarCapture` — orchestrator (996 lines, but manages complex state machine)
- `AnimatedAvatarControls` — filmstrip picker, framing slider, outline toggle
- `AnimatedAvatarBackdropPanel` — color swatch grid
- `AnimatedAvatarCameraPicker` — device selector
- `AnimatedAvatarReviewNav` — section tabs for review phase
- `AvatarCustomColorPanel` — extracted HSV picker (shared with emoji tab)

**Hover playback** — `ProfileAvatar` and `UserAvatar` both parse the animated URL and swap between poster/animation on `mouseEnter`/`mouseLeave`. The `UserProfilePopover` always shows the animation since it's already a hover context.

### Observations

1. **Segmenter singleton** — `segmenterPromise` is module-level, created once per app lifetime. Failed loads clear the promise so the next call retries. `preloadAvatarSegmenter()` warms the model while the user lines up their shot.

2. **Temporal stabilization** — Both the mask confidence (`stabilizeMaskConfidence`) and dark pixel noise (`stabilizeDarkDetailNoise`) use frame-to-frame blending with motion-aware thresholds. This prevents hair shimmer in the looped APNG without smearing real movement.

3. **Composition clip path** — The "pop-out" effect uses a clip region that combines the backdrop circle with a rectangular column above it, so the person can rise above the circle's top edge without spilling past its sides. Clever geometry.

4. **APNG truecolor** — `APNG_COLOR_COUNT = 0` preserves full 24-bit color + 8-bit alpha. The comment explains why: palette quantization caused shimmer in dark areas like hair.

5. **Camera permissions** — `Entitlements.plist` adds `com.apple.security.device.camera` and `Info.plist` adds `NSCameraUsageDescription`. Required for macOS camera access in Tauri.

6. **`upng-js` type declaration** — Clean ambient module declaration in `desktop/src/upng-js.d.ts` since the package doesn't ship types.

7. **Default mode change** — Voice input mode default changed from PTT to VAD in the Rust state (this is also in #1041 — potential merge conflict if they land in different order, but both set the same value).

8. **`haptics.ts` refactor** — `performDefaultHaptic()` extracted as the general-purpose function, `performSidebarDefaultHaptic()` becomes a thin wrapper. Non-breaking.

9. **`AvatarCustomColorPanel` extraction** — The HSV picker was previously inline in `ProfileAvatarEditor`. Extracting it into a shared component allows both the emoji tab and the animated tab to use the same color picker. Good DRY.

10. **`ProfileAvatarEditor` mode expansion** — `AvatarMode` gains `"animated"` as a third tab. The `MODE_TAB_ORDER` array and portal-based preview container make the animated tab's camera feed render in the host's avatar preview slot.

### Potential Issues

1. **`AnimatedAvatarCapture.tsx` at 996 lines** — This is the largest single component. It manages a complex state machine (idle → camera → recording → review → saving) with many UI states. The helpers file extracts constants and utilities, but the component itself could benefit from further decomposition in a follow-up (e.g., extracting the drag/keyboard interaction logic into a custom hook). Not blocking — the state machine is inherently complex and the code is readable.

2. **MediaPipe CDN dependency** — The segmentation model loads from `cdn.jsdelivr.net` and `storage.googleapis.com`. If either is blocked (corporate firewall, China), segmentation silently fails and the avatar records without background removal. This is the correct fallback behavior, but worth documenting for users who expect background removal.

3. **`failedSrc` tracking in `ProfileAvatar`** — The new per-URL failure tracking (`failedSrc === liveSrc`) is more correct than the previous boolean `liveFailed`, but it means a transiently failed poster won't retry until the component remounts or the URL changes. Acceptable for avatar images which rarely transiently fail.

4. **E2e test at 423 lines** — The screenshot spec is thorough (covers camera picker, recording, review, backdrop colors, framing, filmstrip, poster selection, custom color picker, onboarding integration). The fake camera uses a canvas-generated gradient stream, which is a good pattern for deterministic headless testing.

5. **`visibleUrlDraft` removal** — The `ProfileAvatarEditor.utils.ts` change removes `visibleUrlDraft` and the `hiddenAvatarUrl` prop. The URL draft now always starts empty. This simplifies the editor but means re-opening the editor won't pre-populate the URL field with the current avatar URL. This appears intentional — the URL field is for pasting new URLs, not displaying the current one.

### Verdict

Comprehensive, well-architected feature addition. The URL scheme is backward-compatible, the capture pipeline handles edge cases gracefully (offline, no camera, segmentation failure), the component decomposition is reasonable for the complexity, and the test coverage is strong. The DSP work (denoising, mask stabilization, composition) is particularly well-tuned.

**Approve** ✅

**Review by wpfleger96 — APPROVED** [review]

File: desktop/src/features/profile/lib/animatedAvatarCapture.ts [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+  if (!navigator.mediaDevices?.enumerateDevices) {
+    return [];
+  }
+
+  const devices = await navigator.mediaDevices.enumerateDevices();
+  return devices
+    .filter((device) => device.kind === "videoinput" && device.deviceId)
+    .map((device, index) => ({
+      deviceId: device.deviceId,
+      label: device.label || `Camera ${index + 1}`,
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/profile/lib/animatedAvatarCapture.ts]: 🤖 Fixed in c0a86eb7. `listAvatarCameras()` now preserves the browser-provided label verbatim, including blank pre-permission labels, and I added a regression test covering blank videoinput labels so the iPhone/Continuity path stays selectable before permission.

File: desktop/src/features/settings/ui/ProfileSettingsCard.tsx [hunk 2]
```diff
@@ -387,250 +487,352 @@ export function ProfileSettingsCard({
...
+                        data-testid="profile-avatar-editor-shell"
+                        inert={isAvatarEditorOpen ? undefined : true}
+                      >
+                        <ProfileAvatarEditor
+                          animatedPreviewContainer={animatedPreviewEl}
+                          avatarUrl={avatarUrlDraft}
+                          disabled={
+                            updateProfileMutation.isPending &&
+                            !isAvatarEditorFinishing
+                          }
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/settings/ui/ProfileSettingsCard.tsx]: 🤖 Fixed in 3543e3de. `ProfileAvatarEditor` now receives `disabled={isAvatarEditorSaving}`, so the mode tabs and portaled animated preview stay disabled during the pending save and finishing transition. Checks passed: `pnpm exec biome check src/features/settings/ui/ProfileSettingsCard.tsx` and `pnpm typecheck`.

File: desktop/src/shared/lib/animatedAvatar.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  posterUrl: string;
+  /** Animated image played while the avatar is hovered. */
+  animationUrl: string;
+};
+
+export function buildAnimatedAvatarUrl(
+  posterUrl: string,
+  animationUrl: string,
+): string {
+  return `${posterUrl}${ANIMATED_AVATAR_SEPARATOR}${encodeURIComponent(animationUrl)}`;
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/shared/lib/animatedAvatar.ts:28]: 🤖 Fixed in 35f9c9a6. `persistSelfProfile()` now uses `getAvatarSnapshotUrl()` before `rewriteRelayUrl()`, so animated avatar cache fetches request the poster URL while preserving the full `poster#buzz-anim=...` value in storage. Checks passed: `pnpm exec biome check src/shared/lib/animatedAvatar.ts src/shared/lib/animatedAvatar.test.mjs src/features/profile/hooks.ts`, `pnpm typecheck`, and `pnpm test`.

File: desktop/src/features/onboarding/ui/AvatarStep.tsx [hunk 4]
```diff
@@ -237,15 +237,58 @@ export function AvatarStep({ actions, direction, state }: AvatarStepProps) {
...
+    string | null
+  >(null);
+  const [pendingAnimatedAvatarUrl, setPendingAnimatedAvatarUrl] =
+    React.useState<string | null>(null);
   const [isCustomColorPickerOpen, setIsCustomColorPickerOpen] =
     React.useState(false);
-  const canSubmit =
-    avatar.draftUrl.trim().length > 0 && !isSaving && !isUploadingAvatar;
+  const hasAvatarDraft = avatar.draftUrl.trim().length > 0;
+  const canSubmit = hasAvatarDraft && !isSaving && !isUploadingAvatar;
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/onboarding/ui/AvatarStep.tsx:254]: 🤖 Fixed in 57319aa0. `AnimatedAvatarCapture` now reports its internal apply/upload pending state up to the editor, and standalone/onboarding usage forwards that through `onUploadingChange`, so the onboarding Next button stays blocked while the animated avatar upload is in flight. Checks passed: `pnpm exec biome check src/features/profile/ui/AnimatedAvatarCapture.tsx src/features/profile/ui/ProfileAvatarEditor.tsx tests/e2e/animated-avatar-screenshots.spec.ts`, `pnpm typecheck`, `pnpm run build`, `pnpm exec playwright test tests/e2e/animated-avatar-screenshots.spec.ts --project smoke`, and `pnpm test`.

File: desktop/src/features/profile/ui/ProfileAvatarEditor.tsx [hunk 5]
```diff
@@ -449,6 +468,63 @@ export function ProfileAvatarEditor({
...
+    onDone &&
+    !isAnyCustomColorPickerVisible &&
+    (mode !== "animated" || hasAnimatedApply || isDoneButtonPending);
+  const modeTabs = (
+    <Tabs
+      className="w-full"
+      onValueChange={(nextMode) => {
+        if (isInputDisabled) {
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/profile/ui/ProfileAvatarEditor.tsx:481]: 🤖 Fixed in 57319aa0. `isAnimatedApplyPending` now feeds into `isInputDisabled`, so the mode tabs and animated controls are disabled while the settings Done upload is in flight. I also added a smoke E2E assertion that the Image, Emoji, and Animated tabs are disabled during Saving. Checks passed: `pnpm exec biome check src/features/profile/ui/AnimatedAvatarCapture.tsx src/features/profile/ui/ProfileAvatarEditor.tsx tests/e2e/animated-avatar-screenshots.spec.ts`, `pnpm typecheck`, `pnpm run build`, `pnpm exec playwright test tests/e2e/animated-avatar-screenshots.spec.ts --project smoke`, and `pnpm test`.

File: desktop/src/features/onboarding/ui/AvatarStep.tsx [hunk 6]
```diff
@@ -237,15 +237,58 @@ export function AvatarStep({ actions, direction, state }: AvatarStepProps) {
...
+  const isAutoAdvancingAnimatedAvatar = pendingAnimatedAvatarUrl !== null;
+  const shouldHideActionsForAnimatedAvatar =
+    avatarEditorMode === "animated" &&
+    (!hasAvatarDraft || isAutoAdvancingAnimatedAvatar);
```
[the remark this answers is not in the document]
Comment (klopez4212) [reply desktop/src/features/onboarding/ui/AvatarStep.tsx:258]: 🤖 Fixed in 3b81bf30. `AvatarStepActions` now disables Back, Skip for now, and Continue without saving while either the profile save or avatar upload is pending, so an in-flight animated avatar apply cannot be interrupted by onboarding navigation while it later writes the URL. Checks passed: `pnpm check`, `pnpm typecheck`, `pnpm run build`, and `pnpm exec playwright test tests/e2e/onboarding.spec.ts --project integration -g "avatar step"`.
